### PR TITLE
Highlight visited chests

### PR DESCRIPTION
**Pull request by xenonca**
Make it visible for your teammates when a chest has already been opened by someone.
AFAIK it is not ([yet](https://github.com/minetest/minetest/issues/3528)) possible to change node textures during run time, this is why I'm using `minetest.swap_node()`.

### CODE REVIEW

**Comment by LoneWolfHT** [comment]
How are opened chests handled in MTG?

**Comment by savilli** [comment]
> How are opened chests handled in MTG?

The same approach
https://github.com/minetest/minetest_game/blob/42baede13fdf855773cc44ce10f4a3ea4e239404/mods/default/chests.lua#L134

File: mods/ctf/ctf_map/nodes.lua [hunk 1]
```diff
@@ -124,14 +124,17 @@ minetest.register_alias("ctf_map:torch_ceiling", "default:torch_ceiling")
...
+	tiles = {"default_chest_top.png^[colorize:#000000:50^[crack:1:1:1",
+	"default_chest_top.png^[colorize:#000000:50^[crack:1:1:1",
+	"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+	"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+	"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+	"default_chest_front.png^[colorize:#000000:50^[crack:1:1:1"},
 	paramtype2 = "facedir",
 	groups = {immortal = 1},
-	light_source = 2,
+	light_source = 1,
```
Comment (savilli) [inline mods/ctf/ctf_map/nodes.lua:137]: Why?

File: mods/ctf/ctf_map/nodes.lua [hunk 2]
```diff
@@ -172,3 +175,17 @@ minetest.register_node("ctf_map:chest", {
 		end
 	end,
 })
+
+minetest.register_node("ctf_map:chest", {
```
Comment (savilli) [inline mods/ctf/ctf_map/nodes.lua:179]: I'm afraid you need to copy the whole spec, or a closed chest would be diggable, and you could put items inside.

File: mods/ctf/ctf_map/nodes.lua (hunk 1 above)
Comment (xenonca) [reply mods/ctf/ctf_map/nodes.lua:137]: Better distinction in the dark
grafik

**Comment by xenonca** [comment]
> How are opened chests handled in MTG?
I think I should have been more clear with the explanation, what I mean is that visited chests, chests that have already been touched/opened by a player, will be distinguishable from untouched ones.

File: mods/ctf/ctf_map/nodes.lua (hunk 2 above)
Comment (xenonca) [reply mods/ctf/ctf_map/nodes.lua:179]: From what I've tested they are not diggable, not sure how you'd be able to put items inside, the node changes instantly at right-click, is there another way to access the chest?

Comment (savilli) [reply mods/ctf/ctf_map/nodes.lua:179]: I think you can do it with modified client, I wouldn't risk.

Comment (savilli) [reply mods/ctf/ctf_map/nodes.lua:179]: Looks like `can_dig` does nothing and can be removed.

**Comment by appgurueu** [comment]
Looks like massive code duplication to me; shouldn't this basically be the same def. as the treasure chest? Why don't you `table.copy` the treasure chest def. then?

File: mods/ctf/ctf_map/nodes.lua [hunk 3]
```diff
@@ -124,6 +124,52 @@ minetest.register_alias("ctf_map:torch_ceiling", "default:torch_ceiling")
...
+	"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+	"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+	"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+	"default_chest_front.png^[colorize:#000000:50^[crack:1:1:1"},
+	paramtype2 = "facedir",
+	groups = {immortal = 1},
+	light_source = 1,
+	is_ground_content = false,
+	sounds = default.node_sound_wood_defaults(),
+	allow_metadata_inventory_put = function(pos, listname, index, stack, player)
```
Comment (savilli) [inline mods/ctf/ctf_map/nodes.lua:140]: Plz no, too much copy-paste. Look how MTG does and do the same https://github.com/minetest/minetest_game/blob/42baede13fdf855773cc44ce10f4a3ea4e239404/mods/default/chests.lua#L241

**Review by savilli — CHANGES_REQUESTED** [review]

**Review by savilli — APPROVED** [review]
LGTM

File: mods/ctf/ctf_map/nodes.lua [hunk 4]
```diff
@@ -188,4 +188,31 @@ minetest.register_node("ctf_map:chest", {
...
+"default_chest_top.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_front.png^[colorize:#000000:50^[crack:1:1:1"}
+ochest_def.light_source = 1
+ochest_def.on_construct = function(pos)
+	local meta = minetest.get_meta(pos)
+	meta:set_string("infotext", "Treasure Chest (visited)")
+	meta:set_string("formspec", chest_formspec)
```
Comment (appgurueu) [inline mods/ctf/ctf_map/nodes.lua:210]: Couldn't you avoid duplicating this code by calling the original on_construct here?

File: mods/ctf/ctf_map/nodes.lua [hunk 5]
```diff
@@ -188,4 +188,31 @@ minetest.register_node("ctf_map:chest", {
...
+ochest_def.tiles = {"default_chest_top.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_top.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_front.png^[colorize:#000000:50^[crack:1:1:1"}
+ochest_def.light_source = 1
+ochest_def.on_construct = function(pos)
+	local meta = minetest.get_meta(pos)
+	meta:set_string("infotext", "Treasure Chest (visited)")
```
Comment (appgurueu) [inline mods/ctf/ctf_map/nodes.lua:209]: Is duplicating this here even necessary? `swap_node` won't call `on_construct` or `on_destruct`; the previous chest meta will be preserved anyways.

Comment (savilli) [reply mods/ctf/ctf_map/nodes.lua:209]: What if you want to place a visited chest manually?

Comment (appgurueu) [reply mods/ctf/ctf_map/nodes.lua:209]: Why would you want that?

Comment (xenonca) [reply mods/ctf/ctf_map/nodes.lua]: I assume you don't (at least I couldn't think of any situation where you'd want that).

File: mods/ctf/ctf_map/nodes.lua [hunk 6]
```diff
@@ -188,4 +189,23 @@ minetest.register_node("ctf_map:chest", {
 			minetest.show_formspec(player:get_player_name(), "", player:get_inventory_formspec())
 		end
 	end,
-})
+	on_rightclick = function(pos, node, clicker, itemstack, pointed_thing)
+		minetest.swap_node(pos, {name="ctf_map:chest_opened"})
+		local meta = minetest.get_meta(pos)
+		meta:set_string("infotext", chestv)
```
Comment (magnetar47) [inline mods/ctf/ctf_map/nodes.lua:195]:
```suggestion
		minetest.get_meta(pos):set_string("infotext", chestv)
```

No need for a local here unless you're planning to reuse it.

File: mods/ctf/ctf_map/nodes.lua [hunk 7]
```diff
@@ -188,4 +189,23 @@ minetest.register_node("ctf_map:chest", {
 			minetest.show_formspec(player:get_player_name(), "", player:get_inventory_formspec())
 		end
 	end,
-})
+	on_rightclick = function(pos, node, clicker, itemstack, pointed_thing)
+		minetest.swap_node(pos, {name="ctf_map:chest_opened"})
```
Comment (magnetar47) [inline mods/ctf/ctf_map/nodes.lua:193]:
```suggestion
		minetest.swap_node(pos, { name = "ctf_map:chest_opened" })
```

Sry for being pedantic, but better to fix codestyle here itself. :)

File: mods/ctf/ctf_map/nodes.lua [hunk 8]
```diff
@@ -188,4 +189,23 @@ minetest.register_node("ctf_map:chest", {
...
+}
+
+local ochest_def = table.copy(chest_def)
+ochest_def.description = chestv
+ochest_def.tiles = {"default_chest_top.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_top.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
+"default_chest_front.png^[colorize:#000000:50^[crack:1:1:1"}
```
Comment (magnetar47) [inline mods/ctf/ctf_map/nodes.lua:206]:
```suggestion
ochest_def.tiles = {
	"default_chest_top.png^[colorize:#000000:50^[crack:1:1:1",
	"default_chest_top.png^[colorize:#000000:50^[crack:1:1:1",
	"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
	"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
	"default_chest_side.png^[colorize:#000000:50^[crack:1:1:1",
	"default_chest_front.png^[colorize:#000000:50^[crack:1:1:1"}
```

File: mods/ctf/ctf_map/nodes.lua [hunk 9]
```diff
@@ -188,4 +189,23 @@ minetest.register_node("ctf_map:chest", {
...
+ochest_def.light_source = 1
```
Comment (magnetar47) [inline mods/ctf/ctf_map/nodes.lua:207]: What does this do? 👀

Comment (appgurueu) [reply mods/ctf/ctf_map/nodes.lua:207]: It makes the chest glow.

Comment (magnetar47) [reply mods/ctf/ctf_map/nodes.lua:207]: That much I figured. I was wondering about the *why* part. :)

Comment (appgurueu) [reply mods/ctf/ctf_map/nodes.lua:207]: To "highlight" the chest

Comment (xenonca) [reply mods/ctf/ctf_map/nodes.lua:207]: In fact it makes visited chests glow _less_ than the ones which haven't been opened yet.

Comment (magnetar47) [reply mods/ctf/ctf_map/nodes.lua:207]: > In fact it makes visited chests glow _less_ than the ones which haven't been opened yet.

Ah so unopened chests glow even more. Didn't know that.

Comment (magnetar47) [reply mods/ctf/ctf_map/nodes.lua:207]: > In fact it makes visited chests glow _less_ than the ones which haven't been opened yet.

So where's the code that makes unopened chests glow more?

Comment (xenonca) [reply mods/ctf/ctf_map/nodes.lua:207]: https://github.com/MT-CTF/capturetheflag/blob/54cb7a6f67d64f07e68389a22159cf8f1c8a6045/mods/ctf/ctf_map/nodes.lua#L143

File: mods/ctf/ctf_map/nodes.lua (hunk 5 above)
Comment (magnetar47) [reply mods/ctf/ctf_map/nodes.lua]: Is this resolved now?

File: mods/ctf/ctf_map/nodes.lua (hunk 9 above)
Comment (magnetar47) [reply mods/ctf/ctf_map/nodes.lua:207]: Oh I missed it. Thanks.

**Pull request merged**